cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/streaming/function/StreamingMultiSinkFunction.java

```diff
@@ -0,0 +1,194 @@
+import io.cdap.cdap.etl.spark.plugin.SparkPipelinePluginContext;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.function.Function2;
+import org.apache.spark.streaming.Time;
+import org.apache.tephra.TransactionFailureException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Function used to write a batch of data to a batch sink for use with a JavaDStream.
+ * note: not using foreachRDD(VoidFunction2) method, because spark 1.3 doesn't have VoidFunction2.
+ */
+public class StreamingMultiSinkFunction implements Function2<JavaRDD<RecordInfo<Object>>, Time, Void> {
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingMultiSinkFunction.class);
+  private final JavaSparkExecutionContext sec;
+  private final PhaseSpec phaseSpec;
+  private final Set<String> group;
+  private final Set<String> sinkNames;
+  private final Map<String, StageStatisticsCollector> collectors;
+
+  public StreamingMultiSinkFunction(JavaSparkExecutionContext sec, PhaseSpec phaseSpec,
+                                    Set<String> group, Set<String> sinkNames,
+                                    Map<String, StageStatisticsCollector> collectors) {
+    this.sec = sec;
+    this.phaseSpec = phaseSpec;
+    this.group = group;
+    this.sinkNames = sinkNames;
+    this.collectors = collectors;
+  }
+
+  @Override
+  public Void call(JavaRDD<RecordInfo<Object>> data, Time batchTime) throws Exception {
+    long logicalStartTime = batchTime.milliseconds();
+    MacroEvaluator evaluator = new DefaultMacroEvaluator(new BasicArguments(sec),
+                                                         logicalStartTime,
+                                                         sec.getSecureStore(),
+                                                         sec.getNamespace());
+    PluginContext pluginContext = new SparkPipelinePluginContext(sec.getPluginContext(), sec.getMetrics(),
+                                                                 phaseSpec.isStageLoggingEnabled(),
+                                                                 phaseSpec.isProcessTimingEnabled());
+    SparkBatchSinkFactory sinkFactory = new SparkBatchSinkFactory();
+    PipelineRuntime pipelineRuntime = new SparkPipelineRuntime(sec, logicalStartTime);
+
+    Map<String, SubmitterLifecycle<?>> stages = createStages(evaluator);
+
+    // call prepareRun() on all the stages in the group
+    // need to call it in an order that guarantees that inputs are called before outputs
+    // this is because plugins can call getArguments().set() in the prepareRun() method,
+    // which downstream stages should be able to read
+    List<String> traversalOrder = new ArrayList(group.size());
+    for (String stageName : phaseSpec.getPhase().getDag().getTopologicalOrder()) {
+      if (group.contains(stageName)) {
```

Comment:
It seems like `traversalOrder` is only used to affect sinks, is there a way to apply the filtering here instead having it in run method of the TxRunable in the prepareRun and onRunFinish methods? It would help with readability. Can we do `sinkNames.contains(stageName)` instead of this check?

Reply:
it affects transforms in the group as well. We need to call prepareRun() on things in order to make sure that if a transform before a sink sets an argument, the sink will be able to read it.